File: src/test/java/net/dontdrinkandroot/gitki/model/PathTest.java
package net.dontdrinkandroot.gitki.model;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class PathTest
{
    @Test
    public void testDirectoryPath() {
        DirectoryPath path;

        path = new DirectoryPath();
        Assertions.assertTrue(path.isRoot());
        Assertions.assertEquals("", path.toString());

        path = new DirectoryPath("asdf");
        Assertions.assertFalse(path.isRoot());
        Assertions.assertEquals("asdf/", path.toString());

        path = DirectoryPath.parse("asdf/qwer/");
        Assertions.assertFalse(path.isRoot());
        Assertions.assertEquals("asdf/qwer/", path.toString());

        path = path.getParent();
        Assertions.assertEquals("asdf/", path.toString());

        path = path.getParent();
        Assertions.assertEquals("", path.toString());

        path = new DirectoryPath("asdf");
        FilePath filePath = path.appendFileName("qwer");
        Assertions.assertEquals("asdf/qwer", filePath.toString());

        path = new DirectoryPath("asdf");
        path = path.appendDirectoryName("qwer");
        Assertions.assertEquals("asdf/qwer/", path.toString());
    }

    @Test
    public void testToPath()
    {
        FilePath filePath;
        DirectoryPath directoryPath;

        FilePath path = new FilePath("asdf/qwer");
        Assertions.assertEquals("asdf/qwer", path.toPath().toString());

        directoryPath = new DirectoryPath("asdf/qwer/");
        Assertions.assertEquals("asdf/qwer", path.toPath().toString());
    }

    @Test
    public void testFilePath() {
        FilePath filePath;
        DirectoryPath directoryPath;

        filePath = FilePath.parse("asdf");
        Assertions.assertEquals("asdf", filePath.getName());
        Assertions.assertEquals("asdf", filePath.toString());

        directoryPath = filePath.getParent();
        Assertions.assertTrue(directoryPath.isRoot());

        filePath = FilePath.parse("asdf/qwer");
        Assertions.assertEquals("qwer", filePath.getName());
        Assertions.assertEquals("asdf/qwer", filePath.toString());

        directoryPath = filePath.getParent();
        Assertions.assertEquals("asdf/", directoryPath.toString());
        Assertions.assertFalse(directoryPath.isRoot());

        directoryPath = directoryPath.getParent();
        Assertions.assertTrue(directoryPath.isRoot());
    }

    @Test
    public void testEquals() {
        Assertions.assertTrue(new DirectoryPath().equals(new DirectoryPath()));
        Assertions.assertTrue(DirectoryPath.parse("asdf/qwer/").equals(DirectoryPath.parse("asdf/qwer/")));
        Assertions.assertTrue(FilePath.parse("asdf/qwer").equals(FilePath.parse("asdf/qwer")));
    }
}

File: src/test/kotlin/net/dontdrinkandroot/gitki/model/PathTest.kt
package net.dontdrinkandroot.gitki.model

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class PathTest {

    @Test
    fun testDirectoryPath() {

        var path: DirectoryPath?

        path = DirectoryPath()
        Assertions.assertTrue(path.root)
        Assertions.assertEquals("", path.toString())

        path = DirectoryPath("asdf")
        Assertions.assertFalse(path.root)
        Assertions.assertEquals("asdf/", path.toString())

        path = DirectoryPath.parse("asdf/qwer/")
        Assertions.assertFalse(path.root)
        Assertions.assertEquals("asdf/qwer/", path.toString())

        path = path.parent
        Assertions.assertEquals("asdf/", path.toString())

        path = path!!.parent
        Assertions.assertEquals("", path.toString())

        path = DirectoryPath("asdf")
        val filePath = path.appendFileName("qwer")
        Assertions.assertEquals("asdf/qwer", filePath.toString())

        path = DirectoryPath("asdf")
        path = path.appendDirectoryName("qwer")
        Assertions.assertEquals("asdf/qwer/", path.toString())
    }

    @Test
    fun testToPath() {
        val filePath = FilePath("asdf/qwer")
        Assertions.assertEquals("asdf/qwer", filePath.nioPath.toString())

        val directoryPath = DirectoryPath("asdf/qwer/")
        Assertions.assertEquals("asdf/qwer", directoryPath.nioPath.toString())
    }

    @Test
    fun testFilePath() {
        var directoryPath: DirectoryPath?
        var filePath: FilePath = FilePath.parse("asdf")
        Assertions.assertEquals("asdf", filePath.name)
        Assertions.assertEquals("asdf", filePath.toString())

        directoryPath = filePath.parent
        Assertions.assertTrue(directoryPath!!.root)

        filePath = FilePath.parse("asdf/qwer")
        Assertions.assertEquals("qwer", filePath.name)
        Assertions.assertEquals("asdf/qwer", filePath.toString())

        directoryPath = filePath.parent
        Assertions.assertEquals("asdf/", directoryPath.toString())
        Assertions.assertFalse(directoryPath!!.root)

        directoryPath = directoryPath.parent
        Assertions.assertTrue(directoryPath!!.root)
    }

    @Test
    fun testEquals() {
        Assertions.assertTrue(DirectoryPath() == DirectoryPath())
        Assertions.assertTrue(DirectoryPath.parse("asdf/qwer/") == DirectoryPath.parse("asdf/qwer/"))
        Assertions.assertTrue(FilePath.parse("asdf/qwer") == FilePath.parse("asdf/qwer"))
    }
}